File: Algorithm/Code_Materials/c_or_cpp/charpter6_tree_and_level/level2/level_order.cpp
#include <iostream>
#include <vector>
#include <queue>

// 分层打印
using namespace std;

struct TreeNode {
    int val;
    TreeNode* left;
    TreeNode* right;
    TreeNode(int x) : val(x), left(NULL), right(NULL) {}
};

vector<vector<int> > levelOrder(TreeNode* root) {
    vector<vector<int> > ret;
    if (!root) {
        return ret;
    }

    queue<TreeNode*> q;
    q.push(root);
    while (!q.empty()) {
        int currentLevelSize = q.size();
        ret.push_back(vector<int>());
        for (int i = 1; i <= currentLevelSize; ++i) {
            auto node = q.front(); q.pop();
            ret.back().push_back(node->val);
            if (node->left) q.push(node->left);
            if (node->right) q.push(node->right);
        }
    }
    return ret;
}



int main() {
    // 创建二叉树，这里以一个示例二叉树为例，具体实现可以根据实际情况进行修改。
    TreeNode* root = new TreeNode(3);
    root->left = new TreeNode(9);
    root->right = new TreeNode(20);
    root->right->left = new TreeNode(15);
    root->right->right = new TreeNode(7);

    // 调用levelOrder函数获取并输出二叉树的层序遍历结果。
    vector<vector<int> > res = levelOrder(root);
    for (auto& row : res) {
        for (auto& val : row) {
            cout << val << " ";
        }
        cout << endl;
    }
    return 0;
}

File: Algorithm/Code_Materials/python_source/charpter6_tree_and_level/level2/level_order.py
# coding=utf-8
from python_source.tool.Tree import Tree

class LevelOrder(object):
    # 将所有层输出到一行
    def basic_level_order(self, root):
        if not root:
            return []
        res = []
        queue = [root]
        while queue:
            # 获取当前队列的长度，这个长度相当于 当前这一层的节点个数
            size = len(queue)
            tmp = []
            # 将队列中的元素都拿出来(也就是获取这一层的节点)，放到临时list中
            # 如果节点的左/右子树不为空，也放入队列中
            # for _ in xrange(size):
            r = queue.pop(0)
            res.append(r.val)
            if r.left:
                queue.append(r.left)
            if r.right:
                queue.append(r.right)
            # 将临时list加入最终返回结果中
        return res

    # 不同的层分开打印
    def level_order(self, root):
        if not root:
            return []
        res = []
        queue = [root]
        while queue:
            # 获取当前队列的长度，这个长度相当于 当前这一层的节点个数
            size = len(queue)
            tmp = []
            # 将队列中的元素都拿出来(也就是获取这一层的节点)，放到临时list中
            # 如果节点的左/右子树不为空，也放入队列中
            for _ in xrange(size):
                r = queue.pop(0)
                tmp.append(r.val)
                if r.left:
                    queue.append(r.left)
                if r.right:
                    queue.append(r.right)
            # 将临时list加入最终返回结果中
            res.append(tmp)
        return res


if __name__ == "__main__":
    tree = Tree()
    root = tree.init_tree()
    levelOrder = LevelOrder()
    res = levelOrder.basic_level_order(root)
    print res
